### ISSUE TITLE
Any chance to use properties instead of fields for Record<T> ?

### ISSUE THREAD

**Issue by bradphelan** [louthy/language-ext#262]
WPF doesn't like binding to fields.

**Comment by louthy** [comment]
Yes, it works with properties as long as they're field backed properties.

**Comment by bradphelan** [comment]
But how about the ordering. Is it garunteed that auto backed properties
fields are defined in the same order as the properties? If not this could
cause some confusion for IComparable and properties.

**Comment by louthy** [comment]
@bradphelan The ordering is done first with fields then properties.  This is the function called to enumerate the members, so you could try it on your types to see how it will come out.  

```c#
        public static IEnumerable<FieldInfo> GetPublicInstanceFields<A>(params Type[] excludeAttrs)
        {
            var excludeAttrsSet = toSet(excludeAttrs.Map(a => a.Name));
            var publicFields = typeof(A)
                .GetTypeInfo()
                .GetAllFields()
                .Where(f =>
                {
                    if (!f.IsPublic || f.IsStatic) return false;
                    if (toSet(f.CustomAttributes.Map(a => a.AttributeType.Name)).Intersect(excludeAttrsSet).Any()) return false;
                    return true;
                });

            var publicPropNames = typeof(A)
                                    .GetTypeInfo()
                                    .GetAllProperties()
                                    .Where(p => p.CanRead && p.GetMethod.IsPublic)
                                    .Where(p => !toSet(p.CustomAttributes.Map(a => a.AttributeType.Name)).Intersect(excludeAttrsSet).Any())
                                    .ToArray();

            var backingFields = typeof(A)
                                    .GetTypeInfo()
                                    .GetAllFields()
                                    .Where(f => f.IsPrivate &&
                                                publicPropNames.Exists(p => f.Name.StartsWith($"<{p.Name}>")))
                                    .ToArray();

            return Enumerable.Concat(publicFields, backingFields);
        }
```
You could also use `[OptOutOfOrd]` to remove fields or properties from the ordering comparison .operation.

**Comment by bradphelan** [comment]
If you look at the doc for [GetFields ](https://msdn.microsoft.com/en-us/library/ch9714z3(v=vs.110).aspx) you will see

> The GetFields method does not return fields in a particular order, such as alphabetical or declaration order. Your code must not depend on the order in which fields are returned, because that order varies.

Though your code doesn't explicity use ``GetFields`` but rather [DeclaredFields](https://msdn.microsoft.com/en-us/library/system.reflection.typeinfo.declaredfields(v=vs.110).aspx) I assume the same caveat with regards to ordering applies. 

There seems to be some discussion and maybe a solution on getting a stable order here

https://stackoverflow.com/questions/8067493/if-getfields-doesnt-guarantee-order-how-does-layoutkind-sequential-work

**Comment by louthy** [comment]
@bradphelan Thanks for that Brad!  That's very useful, and has been rolled into `v2.1.20`.  Note, if you're using NETCore then you will need to migrate to `netstandard 2.0` as `MetadataToken` is not available on `netstandard 1.3`, so I had to update the whole project.  It was in the post anyway, as `netstandard 1.3` has gotten in the way a number of times.

https://github.com/louthy/language-ext/releases/tag/2.1.20

**Comment by bradphelan** [comment]
Glad to be of help. The project is great :)
